ta/pkcs11/src/pkcs11_attributes.c

```diff
@@ -2258,19 +2412,122 @@ static enum pkcs11_rc alloc_copy_attribute_value(struct obj_attrs *head,
 	return PKCS11_CKR_OK;
 }
 
+static enum pkcs11_rc
+encode_rsa_private_key_der(struct obj_attrs *head, void **data, uint32_t *sz)
+{
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+	int i = 0;
+	int mbedtls_rc = 0;
+	int start = 0;
+	int der_size = 0;
+	void *n = NULL;
+	void *p = NULL;
+	void *q = NULL;
+	void *d = NULL;
+	void *e = NULL;
+	uint32_t n_len = 0;
+	uint32_t p_len = 0;
+	uint32_t q_len = 0;
+	uint32_t d_len = 0;
+	uint32_t e_len = 0;
+	uint8_t *buffer = NULL;
+	mbedtls_pk_context pk = { };
+	mbedtls_rsa_context *rsa = NULL;
+	const mbedtls_pk_info_t *pk_info = NULL;
+
+	mbedtls_pk_init(&pk);
+	pk_info = mbedtls_pk_info_from_type(MBEDTLS_PK_RSA);
+	if (mbedtls_pk_setup(&pk, pk_info)) {
+		rc = PKCS11_CKR_GENERAL_ERROR;
+		goto out;
+	}
+
+	rc = get_attribute_ptr(head, PKCS11_CKA_MODULUS, &n, &n_len);
+	if (rc)
+		goto out;
+
+	rc = get_attribute_ptr(head, PKCS11_CKA_PRIME_1, &p, &p_len);
+	if (rc)
+		goto out;
+
+	rc = get_attribute_ptr(head, PKCS11_CKA_PRIME_2, &q, &q_len);
+	if (rc)
+		goto out;
+
+	rc = get_attribute_ptr(head, PKCS11_CKA_PRIVATE_EXPONENT, &d, &d_len);
+	if (rc)
+		goto out;
+
+	rc = get_attribute_ptr(head, PKCS11_CKA_PUBLIC_EXPONENT, &e, &e_len);
+	if (rc)
+		goto out;
+
+	rsa = mbedtls_pk_rsa(pk);
+	mbedtls_rc = mbedtls_rsa_import_raw(rsa, n, n_len, p, p_len,
+					    q, q_len, d, d_len, e, e_len);
+	if (mbedtls_rc) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	if (mbedtls_rsa_complete(rsa)) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	if (mbedtls_rsa_check_privkey(rsa)) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	der_size = n_len * 8;
+	buffer = TEE_Malloc(der_size, TEE_USER_MEM_HINT_NO_FILL_ZERO);
+	if (!buffer) {
+		rc = PKCS11_CKR_DEVICE_MEMORY;
+		goto out;
+	}
+
+	mbedtls_rc = mbedtls_pk_write_key_der(&pk, buffer, der_size);
+	if (mbedtls_rc < 0) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	start = der_size - mbedtls_rc;
+	for (i = 0; i < mbedtls_rc; i++) {
+		buffer[i] = buffer[i + start];
+		buffer[i + start] = 0;
+	}
+
+	*data = buffer;
+	*sz = mbedtls_rc;
+out:
+	mbedtls_pk_free(&pk);
+
+	if (rc)
+		TEE_Free(buffer);
+
+	return rc;
+}
+
 enum pkcs11_rc alloc_key_data_to_wrap(struct obj_attrs *head, void **data,
 				      uint32_t *sz)
 {
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+
 	switch (get_class(head)) {
 	case PKCS11_CKO_SECRET_KEY:
-		if (alloc_copy_attribute_value(head, data, sz))
-			return PKCS11_CKR_ARGUMENTS_BAD;
+		rc = alloc_copy_attribute_value(head, data, sz);
+		break;
```

Comment:
Actaully this fixes an issue in previous commit "ta: pkcs11: Copy secret key data for wrapping" which returned ARGS_BAD instead of `rc`. Yet, I think it's not probelmatic and ok as is. My A-b stands.

Reply:
Agree, no need to return ARGS_BAD here,
because this check is done already in alloc_copy_attribute_value.
remove repetitive check in "ta: pkcs11: Copy secret key data for wrapping"